### PR TITLE
fix(gstd): Allow mutex/read/write lock guard access by holder only

### PR DESCRIPTION
**Pull request by DennisInSky**
At the moment it is allowed to obtain mx/read/write lock guard in the context of one message and use it in the context of another which is wrong. This PR fixes this issue.

@reviewer-or-team

### CODE REVIEW

File: gstd/src/lock/mutex.rs [hunk 1]
```diff
@@ -117,27 +116,59 @@ impl<T> Mutex<T> {
...
 impl<'a, T> Drop for MutexGuard<'a, T> {
     fn drop(&mut self) {
+        self.ensure_access_by_holder();
         unsafe {
-            *self.mutex.locked.get() = None;
-            if let Some(message_id) = self.mutex.queue.dequeue() {
-                crate::exec::wake(message_id).expect("Failed to wake the message");
+            let locked_by = &mut *self.mutex.locked.get();
+            if let Some(owner_msg_id) = *locked_by {
+                if owner_msg_id != self.holder_msg_id {
```
Comment (breathx) [inline gstd/src/lock/mutex.rs:141]: Seems you already have such check 4 lines above

File: gstd/src/lock/rwlock.rs [hunk 2]
```diff
@@ -198,52 +214,87 @@ impl<T> Deref for RwLockReadGuard<'_, T> {
...
     fn drop(&mut self) {
+        self.ensure_access_by_holder();
         unsafe {
-            *self.lock.locked.get() = None;
-            if let Some(message_id) = self.lock.queue.dequeue() {
-                crate::exec::wake(message_id).expect("Failed to wake the message");
+            let locked_by = &mut *self.lock.locked.get();
+            if let Some(owner_msg_id) = *locked_by {
+                if owner_msg_id != self.holder_msg_id {
+                    panic!(
```
Comment (breathx) [inline gstd/src/lock/rwlock.rs:252]: same

File: examples/binaries/waiter/tests/utils.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+use gstd::errors::{ErrorReplyReason, ReplyCode, SimpleExecutionError};
+use gtest::RunResult;
+
+pub const USER_ID: u64 = 10;
+
+#[track_caller]
+pub fn assert_paniced(result: &RunResult, panic_msg: &str) {
```
Comment (breathx) [inline examples/binaries/waiter/tests/utils.rs:7]:
```suggestion
pub fn assert_panicked(result: &RunResult, panic_msg: &str) {
```

File: examples/binaries/waiter/tests/utils.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        result.log()[0].reply_code(),
+        Some(ReplyCode::Error(ErrorReplyReason::Execution(
+            SimpleExecutionError::UserspacePanic
+        )))
+    ));
+    let payload = String::from_utf8(result.log()[0].payload().into())
+        .expect("Unable to decode panic message")
+        .split(',')
+        .map(String::from)
+        .next()
```
Comment (breathx) [inline examples/binaries/waiter/tests/utils.rs:19]:
```suggestion
        .first()
        .map(String::from)
```

File: examples/binaries/waiter/tests/rw_lock_access.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            lock_msg_id,
+            lock_access_result.sent_message_id()
+        ),
+    );
+}
+
+#[test]
+fn deref_mut_w_lock_guard_from_different_msg_fails() {
+    let system = System::new();
+    let (program, lock_msg_id) = init_fixture(&system, RwLockType::Write);
```
Comment (breathx) [inline examples/binaries/waiter/tests/rw_lock_access.rs:153]: all the file seems so copy-pasty so I'd introduce common fn for them and reuse for each tests with 2 args: Command and panic message

File: examples/binaries/waiter/tests/rw_lock_access.rs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        &format!(
+            "Write lock guard held by message {} is being accessed by message {}",
+            lock_msg_id,
+            lock_access_result.sent_message_id()
+        ),
+    );
+}
+
+fn init_fixture(system: &System, lock_type: RwLockType) -> (Program<'_>, MessageId) {
+    system.init_logger_with_default_filter("");
```
Comment (breathx) [inline examples/binaries/waiter/tests/rw_lock_access.rs:171]: why?

File: examples/binaries/waiter/tests/mx_lock_access.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
+use demo_waiter::{Command, LockContinuation, LockStaticAccessSubcommand, MxLockContinuation};
```
Comment (breathx) [inline examples/binaries/waiter/tests/mx_lock_access.rs:1]: same for the file

File: examples/binaries/waiter/src/lib.rs [hunk 8]
```diff
@@ -69,8 +85,8 @@ pub enum RwLockType {
 
 #[derive(Debug, Encode, Decode)]
 pub enum RwLockContinuation {
```
Comment (breathx) [inline examples/binaries/waiter/src/lib.rs:87]: What's the purpose of this to be enum?

File: examples/binaries/waiter/src/lib.rs [hunk 9]
```diff
@@ -56,9 +56,25 @@ pub enum SleepForWaitType {
...
 #[derive(Debug, Encode, Decode)]
-pub enum MxLockContinuation {
+pub enum LockContinuation {
     Nothing,
     SleepFor(u32),
+    MoveToStatic,
+}
+
+#[derive(Debug, Encode, Decode)]
+pub enum MxLockContinuation {
```
Comment (breathx) [inline examples/binaries/waiter/src/lib.rs:66]: same here

File: examples/binaries/waiter/Cargo.toml [hunk 10]
```diff
@@ -15,6 +15,9 @@ gstd.workspace = true
 gear-wasm-builder.workspace = true
 
 [dev-dependencies]
+gtest.workspace = true
+gear-core.workspace = true
```
Comment (breathx) [inline examples/binaries/waiter/Cargo.toml:19]: add issue to remove gear-core from here. its irrational to import/link/build crate here just for 1 wrapper across [u8; 32]

**Review by breathx — CHANGES_REQUESTED** [review]
Minor changes from comments

File: gstd/src/lock/mutex.rs (hunk 1 above)
Comment (DennisInSky) [reply gstd/src/lock/mutex.rs:141]: These are different checkes:
- the first one checks the current message is the guard holder
- the second one checks the current message is the lock owner
in total it checks the guard holder is the lock owner

File: examples/binaries/waiter/tests/utils.rs (hunk 4 above)
Comment (DennisInSky) [reply examples/binaries/waiter/tests/utils.rs:19]: .first()
   | |         -^^^^^ method not found in `Split<'_, char>`

File: examples/binaries/waiter/tests/rw_lock_access.rs (hunk 6 above)
Comment (DennisInSky) [reply examples/binaries/waiter/tests/rw_lock_access.rs:171]: For the testing purposes I build the contract with the `debug` feature on which enables log messages. It has turned out, by default `system.init_logger()` enables all debug messages, but I am not keen on seeing them in tests

File: examples/binaries/waiter/src/lib.rs (hunk 8 above)
Comment (DennisInSky) [reply examples/binaries/waiter/src/lib.rs:87]: If you are asking about why I have introduced this wrapping enum, then the answer is that I know that I will need it for the tests I have for the mutex ownership, where for the mutex I will have a continuation like `Lock`, whereas for the RW case there will be RW specific continuation like `Lock(RwLockType)`

File: examples/binaries/waiter/Cargo.toml (hunk 10 above)
Comment (DennisInSky) [reply examples/binaries/waiter/Cargo.toml:19]: Actually this is not about a wrapper, it is about the `MessageId` returned by the `RunResult::sent_message_id()` method. Should we instead add an issue that the method should return the `MessageId` type from the `gstd` crate instead?

File: examples/binaries/waiter/tests/rw_lock_access.rs (hunk 5 above)
Comment (DennisInSky) [reply examples/binaries/waiter/tests/rw_lock_access.rs]: Nice catch. Shame on me I haven't noticed it myself. Addressed

File: examples/binaries/waiter/Cargo.toml (hunk 10 above)
Comment (breathx) [reply examples/binaries/waiter/Cargo.toml:19]: > Actually this is not about a wrapper, it is about the `MessageId` returned by the `RunResult::sent_message_id()` method. Should we instead add an issue that the method should return the `MessageId` type from the `gstd` crate instead?

maybe after introducing gstd-primitives crate

**Review by breathx — APPROVED** [review]

File: gtest/src/system.rs [hunk 11]
```diff
@@ -42,7 +42,11 @@ impl System {
     }
 
     pub fn init_logger(&self) {
-        let _ = Builder::from_env(Env::default().default_filter_or("gwasm=debug"))
+        self.init_logger_with_default_filter("gwasm=debug")
+    }
+
+    pub fn init_logger_with_default_filter<'a>(&self, default_filter: impl Into<Cow<'a, str>>) {
```
Comment (shamilsan) [inline gtest/src/system.rs:48]: Isn't it just `filter`, not default?
```suggestion
    pub fn init_logger_with_filter<'a>(&self, filter: impl Into<Cow<'a, str>>) {
```

File: gstd/src/lock/mutex.rs [hunk 12]
```diff
@@ -117,27 +116,59 @@ impl<T> Mutex<T> {
...
 
 impl<'a, T> Drop for MutexGuard<'a, T> {
     fn drop(&mut self) {
+        self.ensure_access_by_holder();
         unsafe {
-            *self.mutex.locked.get() = None;
-            if let Some(message_id) = self.mutex.queue.dequeue() {
-                crate::exec::wake(message_id).expect("Failed to wake the message");
+            let locked_by = &mut *self.mutex.locked.get();
+            if let Some(owner_msg_id) = *locked_by {
```
Comment (shamilsan) [inline gstd/src/lock/mutex.rs:140]: As for me `if let Some(a) = b { ... } else { panic!("...") }` can be replaced by `a = b.unwrap_or_else(|| panic!("..."))`

**Review by shamilsan — APPROVED** [review]
Awesome 🔥

File: gtest/src/system.rs (hunk 11 above)
Comment (DennisInSky) [reply gtest/src/system.rs:48]: Not sure this is a proper name for the func. This filter is taken into account only in cases when it is not specified by other means. For the underlaying method this param is called `default_filter`.
What do you think about naming it `init_logger_with_fallback_filter`?

**Pull request merged**